agent/engine/execcmdagent/manager_linux.go

```diff
@@ -0,0 +1,209 @@
+// +build linux
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package execcmdagent
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+)
+
+// RestartIfStopped restarts the ExecCommandAgent in the container passed as parameter, only for ExecCommandAgent-enabled tasks.
+// The status of the ExecCommandAgent in the container is retrieved using a docker exec inspect call, using the dockerExecID
+// stored in container.GetExecCommandAgentMetadata().DockerExecID.
+//
+// If container.GetExecCommandAgentMetadata() is nil, or the ExecCommandAgent is still running, no action is taken.
+// To actually restart the ExecCommandAgent, this function invokes this instance's Start method.
+func (m *manager) RestartIfStopped(ctx context.Context, client dockerapi.DockerClient, task *apitask.Task, container *apicontainer.Container, containerId string) (bool, error) {
+	if !task.IsExecCommandAgentEnabled() {
+		return false, nil
+	}
+	execMD := container.GetExecCommandAgentMetadata()
+	if execMD == nil { // This means m.Start has never been invoked
+		return false, nil
+	}
+	resp, err := client.InspectContainerExec(ctx, execMD.DockerExecID, dockerclient.ContainerExecInspectTimeout)
+	if err != nil {
+		if _, ok := err.(*dockerapi.DockerTimeoutError); ok {
+			return false, err
+		}
+		// TODO: [ecs-exec] consider retry/adding an extra docker exec inspect API call to get more info
+		return false, err
```

Comment:
if this is a transient docker error, then returning an error triggering a submit status change as agent stopped may not make sense (exec agent could still be running). feels like unless we know for sure agent is not running, we should not trigger submit status change. same for docker timeout errors.

Reply:
I see your point. I still want to signal something is wrong in this case. Will think more about it.

Reply:
Ended up adding retries for the inspect. If the inspect fails after all retries, then I return the error, I think it's reasonable in this case to report something is wrong since we could not inspect the status of the ExecCommandAgent, most likely something is wrong.

Reply:
what would happen if it just keeps timing out?

Reply:
Now returning a `RestartStatus` type to indicate if the exec agent was `Restarted`, `NotRestarted`, `Unknown`.
In case of inspect errors, I return `(Unknown, nil)` (i.e. no error)

P.S. I retry the InspectContainerExec command, except for timeout errors, as discussed.